usr/lib/common/sess_mgr.c

```diff
@@ -566,22 +568,29 @@ CK_RV session_mgr_get_op_state(SESSION *sess,
             return CKR_STATE_UNSAVEABLE;
         }
         active_ops++;
-        if (op_data != NULL) {
-            TRACE_ERROR("%s\n", ock_err(ERR_STATE_UNSAVEABLE));
-            return CKR_STATE_UNSAVEABLE;
-        }
         op_data_len = sizeof(OP_STATE_DATA) +
             sizeof(ENCR_DECR_CONTEXT) +
             sess->encr_ctx.context_len + sess->encr_ctx.mech.ulParameterLen;
+        all_data_len += op_data_len;
 
         if (length_only == FALSE) {
             op_data = (OP_STATE_DATA *) data;
 
-            if (*data_len < op_data_len) {
+            if (max_data_len < op_data_len) {
                 TRACE_ERROR("%s\n", ock_err(ERR_BUFFER_TOO_SMALL));
                  return CKR_BUFFER_TOO_SMALL;
             }
 
+            memset(op_data, 0, sizeof(*op_data));
+#ifdef PACKAGE_VERSION
+            strncpy((char *)op_data->library_version, PACKAGE_VERSION,
+                    sizeof(op_data->library_version));
```

Comment:
If the size of `PACKAGE_VERSION` is greater than 16 (`sizeof(op_data->library_version)`), the string in `op_data->library_version` will not be \0 terminated.

Change the string copy here and in all other places to:
```
strncpy((char *)op_data->library_version, PACKAGE_VERSION, sizeof(op_data->library_version) - 1);
```
It is not required to explicitly set the terminating char, because the struct has been initialized in the line before.

Reply:
Well yes, I also thought about this, but it actually does not matter if it is zero terminated or not. 
In `session_mgr_set_op_state()` it is compared using `strncmp`, so zero termination does not matter:
```
if (strncmp((char *)op_data->library_version, PACKAGE_VERSION,
                sizeof(op_data->library_version)) != 0) {
```
With your suggestion, I would also have to change the strncmp() call to use `sizeof(...) - 1` I guess, otherwise it would not match if it has been truncated.

Reply:
`strncmp()` skip the termination character implicitly. There is no need to change these calls.

Reply:
Here is an example:
```
$ cat xxx.c
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

int main(void)
{
        char dest[16];
        char *src_short = "abc";
        char *src_long = "abcdefghijklmnopqrstuvwxyz";

        memset(dest, 0, sizeof(dest));
        printf("%s %ld %ld\n", dest, sizeof(dest), strlen(dest));

        strncpy(dest, src_short, sizeof(dest));
        printf("%s %ld %ld %d\n", dest, sizeof(dest), strlen(dest),
               strncmp(dest, src_short, sizeof(dest)));

        strncpy(dest, src_long, sizeof(dest));
        printf("%s %ld %ld %d\n", dest, sizeof(dest), strlen(dest),
               strncmp(dest, src_long, sizeof(dest)));

        return 0;
}
$ gcc -O0 -g -Wextra -Wall -o xxx xxx.c
$ ./xxx
 16 0
abc 16 3 0
abcdefghijklmno�д�U 16 22 0
```

Reply:
right, but lets assume `PACKAGE_VERSION` is a string of 20 characters. With `op_data->library_version` being truncated to 15 characters (plus 1 zero byte), then the 2 strings will simply not match if `strncmp` is used with a length of 16.

Reply:
They do match, as you can see in my example.

Reply:
But you are using `sizeof(dest)` (which is what I do in my code), not `sizeof(dest) - 1` as you suggested.

Reply:
Ah, sorry. They don't. My fault.

Reply:
so bottom line, I would tend to leave it as is, but maybe add a comment to it that zero termination doesn't matter here.

Reply:
But anyhow, please descide, if `op_data->library_version` is holding a string or not. If it is defined as a string, if must have a \0 termination. Otherwise I would prefer to use `memcpy()` and `memcmp()` instead of the string functions.

Reply:
well, technically ist a CK_CHAR array.... not sure if that makes it a C-string.
The other 2 fields manufacturerID and model are also CK_CHAR arrays and are not zero terminated (but blank padded)....

With memcpy, I would need to copy the C-string PACKAGE_VERSION to it but have to care about its length myself (strlen(PACKAGE_VERSION) - but what of its length is larger then 16....).
Well this is what strncpy is for. Why reinvent the wheel?

I will add a comment to the declaration and leave the code as is.:
`CK_CHAR library_version[16]; /* zero termination does not matter here */`

Reply:
ok, fine with me. Let's keep fingers crossed, that it won't hurt us in the future.

Reply:
Well, PACKAGE_VERSION contains "3.18.0" as of today. Imagine a real live version number that's larger than 15 characters.... ;-)